spec_util/meld.go

```diff
@@ -398,20 +418,44 @@ func assignDataFormats(d *pb.Data, formats map[string]bool) {
 }
 
 func meldStruct(dst, src *pb.Struct) error {
+	// If a field appears in both structs, it is assumed to be required.
+	// If it appears in one, but not the other, then it should become
+	// optional (if not optional already.)
+
 	if dst.Fields == nil {
 		dst.Fields = src.Fields
 		return nil
 	}
+	for k, dstData := range dst.Fields {
+		if _, ok := src.Fields[k]; !ok {
+			// Fields in dst but not in src.
+			makeOptional(dstData)
+		}
+	}
 	for k, srcData := range src.Fields {
 		if dstData, ok := dst.Fields[k]; ok {
+			// Found in both, MeldData handles if either is already
+			// optional.
 			if err := MeldData(dstData, srcData); err != nil {
 				return errors.Wrapf(err, "failed to meld struct key %s", k)
 			}
 		} else {
-			// The meld is additive - any new field is included.
+			// Fields found in src but not in dst.
+			makeOptional(srcData)
 			dst.Fields[k] = srcData
 		}
 	}
+
+	// ... but what if we end up with hundreds of keys?
+	// Probably that means the key is part of the data. Let's use an arbitrary
+	// threshold, and if we cross it, merge all the keys.
+	// TODO: convert to a new Map type instead of a Struct? Very disruptive, have to fix all the visitors etc.
+	// or TODO: use a special key "*" (matching pq syntax?), but somebody could actually use that in practice
+	// or TODO: use a special key that is less likely to be used, but remap it when displaying it or showing
+	// it as an OpenAPI spec.
+	// or TODO: add another field to Struct specifying that this has occurred.
+	// (and if we can infer a data format on the keys, it could live at that level too.)
```

Comment:
I'm leaning toward that last option -- adding another field to Struct indicating that it's a data object, and optionally recording any data formats we can infer.  We could do both keys and values, potentially.  The struct itself could just be empty -- we'll need to test it, but I don't think empty structs will break things.